libcnb/src/layer/test.rs

```diff
@@ -0,0 +1,898 @@
+    "#,
+    )
+    .unwrap();
+
+    let handle_layer_result = handle_layer(&context, layer_name.clone(), test_layer).unwrap();
+
+    // Assert layer content metadata, returned and on disk
+    let expected_content_layer_metadata = LayerContentMetadata {
+        types: LayerTypes {
+            launch: TEST_LAYER_LAUNCH,
+            build: TEST_LAYER_BUILD,
+            cache: TEST_LAYER_CACHE,
+        },
+        metadata: TestLayerMetadata {
+            version: metadata_version_string,
+        },
+    };
+
+    assert_eq!(
+        &handle_layer_result.content_metadata,
+        &expected_content_layer_metadata
+    );
+
+    let layer_content_metadata_from_disk: LayerContentMetadata<TestLayerMetadata> = read_toml_file(
+        temp_dir
+            .path()
+            .join("layers")
+            .join(format!("{}.toml", &layer_name)),
+    )
+    .unwrap();
+
+    // Assert basic LayerData fields
+    assert_eq!(
+        &layer_content_metadata_from_disk,
+        &expected_content_layer_metadata
+    );
+
+    assert_eq!(handle_layer_result.name, layer_name);
+    assert_eq!(
+        handle_layer_result.path,
+        temp_dir.path().join("layers").join(layer_name.as_str())
+    );
+
+    assert_eq!(handle_layer_result.env, LayerEnv::new());
+
+    // Assert layer directory contents
+    let create_file_contents =
+        fs::read_to_string(handle_layer_result.path.join(TEST_LAYER_CREATE_FILE_NAME)).ok();
+
+    let update_file_contents =
+        fs::read_to_string(handle_layer_result.path.join(TEST_LAYER_UPDATE_FILE_NAME)).ok();
+
+    assert_eq!(
+        create_file_contents,
+        Some(String::from(TEST_LAYER_CREATE_FILE_CONTENTS))
+    );
+
+    assert_eq!(update_file_contents, None);
+}
+
+#[test]
+fn error_handling_no_metadata_toml() {
+    let temp_dir = tempdir().unwrap();
+    let context = build_context(&temp_dir);
+    let layer_name = random_layer_name();
+
+    let layer_dir = temp_dir.path().join("layers").join(layer_name.as_str());
+    fs::create_dir_all(&layer_dir).unwrap();
+
+    let handle_layer_result = handle_layer(&context, layer_name, TestLayer::default()).unwrap();
+
+    let create_file_contents =
+        fs::read_to_string(handle_layer_result.path.join(TEST_LAYER_CREATE_FILE_NAME)).ok();
+
+    let update_file_contents =
+        fs::read_to_string(handle_layer_result.path.join(TEST_LAYER_UPDATE_FILE_NAME)).ok();
+
+    assert_eq!(
+        create_file_contents,
+        Some(String::from(TEST_LAYER_CREATE_FILE_CONTENTS))
+    );
+
+    assert_eq!(update_file_contents, None);
+}
+
+#[test]
+fn error_handling_no_directory() {
+    let temp_dir = tempdir().unwrap();
+    let context = build_context(&temp_dir);
+    let layer_name = random_layer_name();
+
+    let layer_toml_path = temp_dir
+        .path()
+        .join("layers")
+        .join(format!("{}.toml", layer_name.as_str()));
+
+    fs::write(
+        &layer_toml_path,
+        r#"
+[types]
+launch = true
+build = true
+cache = true
```

Comment:
I suppose technically this `types` table wouldn't exist here, since the restored layer would have had that stripped by the lifecycle?

Reply:
It currently doesn't matter for this test, but you're totally right! I removed the types in 224f3d4.

Reply:
Thank you :-) I agree it doesn't affect the test, but it both cleans them up and means more representative of actual input.

It can also be removed from `update_with_incompatible_metadata_recreate()` and `update_with_incompatible_metadata_replace()`, but not important if you'd rather not :-)